evmcore/tx_validation.go

```diff
@@ -125,39 +118,76 @@ func validateTx(tx *types.Transaction, opt validationOptions) error {
 		return ErrTipAboveFeeCap
 	}
 
+	// For now, Sonic only supports Blob transactions without blob data.
+	if tx.BlobHashes() != nil && (len(tx.BlobHashes()) > 0 ||
+		(tx.BlobTxSidecar() != nil && len(tx.BlobTxSidecar().BlobHashes()) > 0)) {
+		return ErrTxTypeNotSupported
+	}
+
+	// Check non-empty authorization list
+	if tx.SetCodeAuthorizations() != nil && len(tx.SetCodeAuthorizations()) == 0 {
+		return ErrEmptyAuthorizations
+	}
+
+	return nil
+}
+
+// ValidateTxForNetworkRules runs a set of verifications against the given
+// network rules. It checks the transaction's:
+// - type (legacy, access list, dynamic fee, blob, set code) is supported in the current network
+// - gas is within the max accepted gas per transaction
+// - sender is valid
+// - gas fee cap is greater than the minimum base fee
+// - gas is greater than the intrinsic gas
+// - gas is greater than the floor data cost (EIP-7623)
+//
+// It returns an error if any of the checks fail.
+func ValidateTxForNetworkRules(tx *types.Transaction, opt NetworkRulesForValidateTx) (common.Address, error) {
+
+	// Accept only legacy transactions until EIP-2718/2930 activates.
+	// Since both eip-2718 and eip-2930 are activated in the berlin fork
+	// (https://blog.ethereum.org/2021/03/08/ethereum-berlin-upgrade-announcement),
+	// they can be grouped in a single flag.
+	if !opt.eip2718 && tx.Type() != types.LegacyTxType {
+		return common.Address{}, ErrTxTypeNotSupported
+	}
+	// Reject dynamic fee transactions until EIP-1559 activates.
+	if !opt.eip1559 && tx.Type() == types.DynamicFeeTxType {
+		return common.Address{}, ErrTxTypeNotSupported
+	}
+	// Reject blob transactions until EIP-4844 activates or if is already EIP-4844 and they are not empty
+	if tx.Type() == types.BlobTxType && !opt.eip4844 {
+		return common.Address{}, ErrTxTypeNotSupported
+	}
```

Comment:
If I recall correctly, we do not actually support a version of Sonic where these EIPs are disabled. Does it make sense to have these checks at this fine-granular level or should it be simplified?

@LuisPH3 wdyt?

Reply:
This is something that occurred to me as well, since the earlier revision sonic supports in Cancun, everything before will always be enabled, so this checks could be joined. That said I wanted to keep this PR focus on only splitting the validations and reordering them, as that is a mess of its own, and then add follow-up tasks with refactors.